### PR TITLE
Current Position In Degrees

### PR DESCRIPTION
**Pull request by mitchgres**
I also added the configuration for the angle encoder.

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 1]
```diff
@@ -4,11 +4,17 @@
...
 
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 import frc.robot.hardware.shooter.ShooterHardware;
 
 public class Shooter extends SubsystemBase {
+  // Note: The channel that this encoder is on will need to be configured for the
+  // robot.
+  private final DutyCycleEncoder _angleEncoder = configEncoder(new DutyCycleEncoder(0));
```
Comment (rhit-edgingza) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:15]: We are not creating any hardware inside of subsystems anymore, it interferes with our ability to run tests. Any new piece of hardware for the shooter (this includes motors, sensors, PID controllers, etc) needs to be added to ShooterHardware, RealShooterHardware, and SimShooterHardware. There are examples of how to do this in those files already. The port for the encoder also should be a constant, which goes in HardwareConstants.java, which is also in the hardware folder.

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 2]
```diff
@@ -4,11 +4,17 @@
...
   ShooterHardware _hardware;
+  private static final float ANGLE_ENCODER_OFFSET_FOR_ROBOT_BASE = 0;
```
Comment (rhit-edgingza) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:17]: In an effort to keep subsystems not clogged with constants, each subsystem has its own constants file in its folder. Please put this constant in there. Also, please use double, not float, for the type

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 3]
```diff
@@ -18,4 +24,15 @@ public Shooter(ShooterHardware hardware) {
...
     // This method will be called once per scheduler run
   }
+
+  public double getCurrentPositionInDegrees() {
+    // This assumes that absolute zero is the base of the robot though an offset
+    // might need to be set.
+    return Rotation2d.fromRotations(_angleEncoder.get()).getDegrees();
+  }
+
+  private DutyCycleEncoder configEncoder(DutyCycleEncoder encoder) {
```
Comment (rhit-edgingza) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:34]: Configs will work without you needing to return them. So this method doesn't require an input (just use the getter from _hardware that you'll be adding to get the encoder object) or an output. When you call those methods, the config is applied to the object that we already have access to, so there's no reason to return it.

**Review by rhit-edgingza — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 4]
```diff
@@ -18,4 +21,15 @@ public Shooter(ShooterHardware hardware) {
   public void periodic() {
     // This method will be called once per scheduler run
   }
+
+  public double getCurrentPositionInDegrees() throws RuntimeException {
+    double encoderValueAsRotations = _hardware.getAngleEncoder().get();
+    if (encoderValueAsRotations > ShooterConstants.MAXIMUM_ANGLE_ENCODER_TURNS
+        || encoderValueAsRotations < ShooterConstants.MINIMUM_ANGLE_ENCODER_TURNS) {
+      throw new RuntimeException(
+          "It's impossible for the encoder to be this value. There must be a hardware error. Shut down this subsystem to not break everything.");
```
Comment (rhit-edgingza) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:30]: We don't want to do this in this way. The stops we have here are soft stops, in all reality, a PID controlling this motor may go slightly beyond these stops as it fluctuates, which is ok since we set soft stops well before hard stops where things start to break. So throwing an exception here to stop the code will happen quite frequently if we have a setpoint near these boundaries. If you really want to have this sort of mechanism in place, add in additional stops that are 5~10 degrees beyond these stops that would cause the code to crash

File: src/main/java/frc/robot/subsystems/shooter/Shooter.java [hunk 5]
```diff
@@ -18,4 +21,15 @@ public Shooter(ShooterHardware hardware) {
...
+    } else {
+      return Rotation2d.fromRotations(encoderValueAsRotations).getDegrees();
```
Comment (rhit-edgingza) [inline src/main/java/frc/robot/subsystems/shooter/Shooter.java:32]: We don't want to do a conversion like this between rotations and degrees because we have no control over what constitutes 0. In this case, 0 would correspond to the 0 position of the encoder, but recall that our 0 is likely to be the position of the shooter when it is parallel to the ground. We should approach this similarly to last year, where we had two constants for our max and our min positions in degrees with their values being the value of the encoder at that position, then we figured out the ratio of encoder outputs to degrees by dividing the difference between the two positions in degrees by the difference between the two positions in encoder output. Let's have that calculation be a constant that gets calculated at the top of this class and then to return the encoder position in degrees you can multiply the encoder output by that ratio.

File: src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package frc.robot.subsystems.shooter;
+
+public abstract class ShooterConstants {
+    public static final double MAXIMUM_ANGLE_ENCODER_TURNS = 10;
```
Comment (rhit-edgingza) [inline src/main/java/frc/robot/subsystems/shooter/ShooterConstants.java:4]: Absolute encoders aren't like potentiometers, we only get one rotation before they roll over. Instead, have these be in degrees, which will make more sense to people when they use them

File: src/test/java/shooter/GetCurrentPositionInDegreesTests.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package shooter;
```
Comment (rhit-edgingza) [inline src/test/java/shooter/GetCurrentPositionInDegreesTests.java:1]: Only keep these tests if you add the extra stop constants I mentioned above. If you do, you are missing a case based on BVA. You should test right at the boundaries on both sides, so the first test should be testing one boundary and you should add one more test to test the other boundary

**Review by rhit-edgingza — CHANGES_REQUESTED** [review]

File: src/test/java/shooter/GetCurrentPositionInDegreesTests.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @SuppressWarnings("PMD.SignatureDeclareThrowsException")
+    @AfterEach
+    void shutdown() throws Exception {
+
+    }
+
+    @Test
+    void testGetCurrentPositionInDegrees_withTurnsOutput_expectDegrees() {
```
Comment (rhit-edgingza) [inline src/test/java/shooter/GetCurrentPositionInDegreesTests.java:48]: you're still missing a test. You need to check that the bounds that you made for throwing an exception work correctly. please write a test that checks just inside and outside MAXIMUM_ANGLE_ENCODER_TURNS + 0.1 and just inside and outside MINIMUM_ANGLE_ENCODER_TURNS - 0.1

**Review by rhit-edgingza — CHANGES_REQUESTED** [review]

**Review by rhit-edgingza — APPROVED** [review]

**Comment by mitchgres** [comment]
Currently working on resolving these merge conflicts. Shouldn't be too bad, and I'll merge into main whenever it builds.
